Feature/sunlight

carschno opened the pull request:
Follow-up to PR https://github.com/Smithsonian/dpo-voyager/pull/374 to complete issue https://github.com/Dynamic3D-Project/dpo-voyager/issues/10

- [X] add sunlight as directional light
- [X] automatically adjust sunlight position to date, time, timezone, latitude and longitude
- [x] adjust sunlight intensity and color

## Follow-up changes (separate PRs):

- serialize datetime for sunlight in snapshots https://github.com/Dynamic3D-Project/dpo-voyager/pull/21
- add compass https://github.com/Smithsonian/dpo-voyager/pull/403

## Implementation details

<img width="2364" height="1134" alt="Screenshot 2025-11-05 at 16-38-48 Voyager Story v0 56 1-73-g10c76b8f DEV" src="https://github.com/user-attachments/assets/1fc8dd28-5d9e-443e-b7f5-7c24fccc1123" />

- In Explorer, users can edit the date and time, and latitude and longitude in the lights tool. The light position, intensity, and color are adapted accordingly.
- In authoring mode, users can additionally adjust the parameters `intensityFactor` and `sunlightDistance` that are used to fine-tune the sunlight intensity and position.

## TODOs

An overview of issues that have come up during review and discussions compiled from the comments below

- [X] There's an 'unhandled property' warning being thrown when the sun settings are rendered
- [X] Sun color appears to not update the UI widget in Story.
- [X] 'Intensity' is editable even though it appears to not actually be (which makes sense)
- [X] Since DateTime is just for informational purposes, maybe it could use a shorter format?
- ~~[ ] What is the purpose of SunDistance? (-> fix by adding documentation/tooltip?)~~
- [X] Are shadows working? (out of scope of this PR).
- [X] The Compass overlaps the main menu in some cases
- [X] this feature is adding about 60% to the package size for the component
- [X] The elements should wrap like they do with other similar toolbars
- [X] We probably don't need to have disabled parameters take up space in the bar.
- [x] date/time, lat, long, all require text input which is probably not a great user experience on mobile
- [X] The "sun" type is missing from document.schema.json
- [X] issue with parsing the DateTime
- [x] Only date/time control needed in explorer, remove lat/long (and timezone?)
- [x] fix light helper ([comment](https://github.com/Smithsonian/dpo-voyager/pull/391#issuecomment-3813781946))
- [x] fix:  When sun light is added, color and intensity are disabled for all other lights in the scene and remain that way even if the sun is deleted.
- [x] fix: the helper isn't showing up at a reasonable scale

gjcope commented:
Hey @carschno - apologies for the delay. Generally, this seems to work pretty great! A few issues/questions below:

- There's an 'unhandled property' warning being thrown when the sun settings are rendered
- Sun color appears to not update the UI widget in Story.
- 'Intensity' is editable even though it appears to not actually be (which makes sense) so you can set the 'disabled' property in the element definition (same for DateTime)
- Since DateTime is just for informational purposes, maybe it could use a shorter format?
- What is the purpose of SunDistance?
- Are shadows working? They appear to be enabled (at least in the UI) but I'm not able to get a result.
- The Compass overlaps the main menu in some cases
- The Explorer UI for adjusting the sun element seems to have a number of formatting/responsiveness issues. Happy to go into more depth if helpful.

I don't really have an answer for this at the moment, but a higher-level concern is that it appears just this feature is adding about 60% to the package size for the component. That's kind of a lot for something that isn't core.

gjcope commented:
> This indicates an increase from 13M to 15M. An increase of 15% still seems not great, but not critical either, does it?

The difference is getting obfuscated here because you are looking at the whole folder. What we are really concerned about is the production js that will be loaded for a Voyager Explorer end user (voyager-explorer.min.js). With the current sunlight and master branches I get:

master: 1483KB, sunlight: 2580KB or an increase of ~74% for this feature.

My assumption is that it has a lot to with the new dependencies. Is there an opportunity for [tree shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking) here?

carschno commented:
> * The elements should wrap like they do with other similar toolbars (see 'View') so we can better accommodate mobile and smaller width screens.

Should be fixed by 17e14a079d37340013ac7b0c3642a4d7063cc38e

> * We probably don't need to have disabled parameters take up space in the bar. I'm not sure we have another instance of this, so this might be new functionality needed.

From a UX perspective, I think it is good to see the changing values for intensity and color when updating the active parameters. 7179a1248257fcae89fc82a18f056e44e35622f0 moves them to the right so that the active parameters come first.

> * It looks like date/time, lat, long, all require text input which is probably not a great user experience on mobile. I realize it may be necessary for accuracy, but maybe we can think about additional support for dragging, calendar widget, etc. I haven't fully thought through those options yet.

Indeed. While looking into this, I realised that `PropertyNumber#render()` hard-codes `type="text"`, which seems to work against the built-in mechanisms for properly using numbers, especially on mobile. Am I missing the point of doing that?
I am very hesitant to change that because it would affect all instances of `sv-property-number`. On the other hand, the newly introduced longitude and latitude fields seem to be the only places to use this element type in the entire code...

Quick update: date/time actually uses the `datetime-local` input type, which I would expect to give an appropriate mobile experience.

carschno commented:
> Compass overlap:

Resolved in 7e9d6b629cee028b24bb0e83e096642a83a79630 : it moves the compass from the SceneView to the ChromeView, which the right place anyway, I suppose, and sets a wider margin to the left so that the compass remains next to the main menu.

afb8b9b76874b86162fe30829b3aaedc90fec8d1 scales the compass according to the window size.

<img width="622" height="482" alt="image" src="https://github.com/user-attachments/assets/050365c5-ee88-437a-b26e-739e4b5ca52f" />

I suppose there is no perfect solution for very small screens, but this prevents the overlap and makes the compass still usable on small screens.

carschno commented:
Summary of the current status:

1. the format/length issues with `DateTime` are resolved through #401 (already merged into this branch)
1. the shadow issues seem unrelated to the sunlight type (which is just a directional light). My impression is that exposing `receiveShadow` would not solve it, but it the user has to adapt the `ShadowSide` parameter (see https://github.com/Smithsonian/dpo-voyager/pull/391#issuecomment-3627798846)
1. the `SunDistance` is a factor to adapt the sun position to the scale of the model. This might require some background knowledge, but I am not sure how to resolve it. There seems to be no way to add a tooltip or other in-place documentation to the settings of a light, are there?

@gjcope please let me know if you still see any issues that block this PR from being merged.

gjcope commented on source/client/ui/story/SettingsTaskView.ts:
```diff
@@ -98,8 +99,10 @@ export class SettingsTree extends Tree<ITreeNode>
         }
 
-        return html`<sv-property-view .property=${node.property}></sv-property-view>`;
+        const disabled = this.node?.light instanceof CVSunLight && 
```
This seems odd here. Is there a reason this logic is not part of the new component?

gjcope commented:
> 1. the format/length issues with `DateTime` are resolved through [Extract PropertyPercent #401](https://github.com/Smithsonian/dpo-voyager/pull/401) (already merged into this branch)

It makes it a little more challenging to evaluate with this PR intertwined. But seems to work well for DateTime with proposed style tweaks below.

> 2. the shadow issues seem unrelated to the sunlight type (which is just a directional light). My impression is that exposing `receiveShadow` would not solve it, but it the user has to adapt the `ShadowSide` parameter (see [Feature/sunlight #391 (comment)](https://github.com/Smithsonian/dpo-voyager/pull/391#issuecomment-3627798846))

It should require both 'receiveShadow' and the appropriate 'ShadowSide' defined. For 'ShadowSide', it really depends on the geometry that is casting the shadow.

> 3. the `SunDistance` is a factor to adapt the sun position to the scale of the model. This might require some background knowledge, but I am not sure how to resolve it. There seems to be no way to add a tooltip or other in-place documentation to the settings of a light, are there?

Not currently, but that sort of functionality has been discussed. Happy to hear a proposal if you have ideas.

> 
> @gjcope please let me know if you still see any issues that block this PR from being merged.

1. I haven't looked at #401 yet so I assume this is related, but I'm not a fan of the +/- or up/down icons that now appear with number inputs.
2. The precision of automatically generated floats shown in the UI is way too long.
3. Explorer UI issues 
<img width="270" height="585" alt="image" src="https://github.com/user-attachments/assets/086e6c22-fa9b-4647-878f-1d0caa610d20" /><br>
- Latitude and Longitude are not visible on mobile.
- The inputs visible on mobile work well functionally, but the styling of DateTime doesn't indicate that it's interactive to me. Can we provide more contrast?

4. Can you put the compass commits into a different PR? We aren't going to be able to immediately merge a new menu button as we are actively going through a UI redesign and I also think it's separate (but related) functionality.

Do you have use cases from @kgschoueri or others on that end for the front-facing sun UI parameters? We don't have any, so I can't really comment on it, but I do wonder what the value of some of them is for the end user based on the examples I have heard. Not against it, just trying to keep things simple where possible.

carschno commented:
> It makes it a little more challenging to evaluate with this PR intertwined. But seems to work well for DateTime with proposed style tweaks below.

I have reverted the commits from #401. It also made me realize that the date/time input field has been using the appropriate input type (`datetime-local`) anyway. Could you double-check if the issue you saw earlier with the datetime selection persists? Perhaps it was related to the excessive list of time zones (resolved)?

> It should require both 'receiveShadow' and the appropriate 'ShadowSide' defined. For 'ShadowSide', it really depends on the geometry that is casting the shadow.

For a lack of experience, I am a bit puzzled if this means that this issue is specific to the sunlight or not. My understand from https://github.com/Smithsonian/dpo-voyager/pull/391#issuecomment-3627798846 has been that it is not, so I would tackle it outside of this PR.

> > 3. the `SunDistance` is a factor to adapt the sun position to the scale of the model. This might require some background knowledge, but I am not sure how to resolve it. There seems to be no way to add a tooltip or other in-place documentation to the settings of a light, are there?
> 
> Not currently, but that sort of functionality has been discussed. Happy to hear a proposal if you have ideas.

No specific ideas at the moment. After a first glance, adding a respective property for the `Property` seems a good approach. I would cover that separately.

>     * Latitude and Longitude are not visible on mobile.

Fixed by reverting the changes from #401 .

>     * The inputs visible on mobile work well functionally, but the styling of DateTime doesn't indicate that it's interactive to me. Can we provide more contrast?

Done: 71351f28089f60c46a77f8492e0b9576ad1e62fd

>     4. Can you put the compass commits into a different PR? We aren't going to be able to immediately merge a new menu button as we are actively going through a UI redesign and I also think it's separate (but related) functionality.

Done, see https://github.com/Smithsonian/dpo-voyager/pull/403
I hope this makes reviewing simpler.
The compass is required for the sunlight to make sense, so I leave the changes in here, too, ok?

carschno commented on source/client/ui/story/CreateLightMenu.ts:
```diff
@@ -6,15 +6,16 @@ import { focusTrap, getFocusableElements } from "../../utils/focusHelpers";
 @customElement("sv-create-light-menu")
 export default class CreateLightMenu extends Popup {
     protected language: CVLanguageManager = null;
+    protected allowSunLight: boolean = true;
```
It prevents the sunlight to be shown as an option. The reasoning is that there should never be more than one sunlight. Therefore, `CreateLightMenu.show(mainView, language, !sunExists)` is called in `NodeTree.ts`.

gjcope replied:
Is there a technical reason there should never be more than one? I know practically that makes sense, but are we breaking anything by not limiting it?

gjcope commented:
> I have reverted the commits from https://github.com/Smithsonian/dpo-voyager/pull/401. It also made me realize that the date/time input field has been using the appropriate input type (datetime-local) anyway. Could you double-check if the issue you saw earlier with the datetime selection persists? Perhaps it was related to the excessive list of time zones (resolved)?

Yes, looks like the date/time field works as expected.

> For a lack of experience, I am a bit puzzled if this means that this issue is specific to the sunlight or not. My understand from https://github.com/Smithsonian/dpo-voyager/pull/391#issuecomment-3627798846 has been that it is not, so I would tackle it outside of this PR.

Yes, you are correct. The comment was just for clarification, it is not specific to sunlight.

>> * Latitude and Longitude are not visible on mobile.
> Fixed by reverting the changes from https://github.com/Smithsonian/dpo-voyager/pull/401 .

This is still not working for me. Only seeing the first two inputs on mobile. We might also consider making the lat/long input styles more visible in Explorer (don't want to affect the style in Story). They also seem a little consistent in that they don't allow dragging?

> The compass is required for the sunlight to make sense, so I leave the changes in here, too, ok?

It's up to you but I suggested separating them since it's currently a blocker to merging this PR due to the UI redesign described above. If you're fine with waiting then no worries. It's also not clear to me that sunlight doesn't make sense without it. But this ties into 
> Do you have use cases from @kgschoueri or others on that end for the front-facing sun UI parameters?

gjcope commented:
> Hey, sorry just seeing this now. I'm not sure I 100% understand what front-facing sun UI parameters mean in this context. @gjcope could you explain a bit what the functionality entails or how it differs from another option your considering? Then I can think if any of our editions might relate.

Hi @kgschoueri - The lights all have configurable parameters in Voyager Story so as the scene author you can set them up to have the specific effect you want to present to the end user. Then in Voyager Explorer (front-facing) we expose a small subset of those parameters (generally just intensity and color) to support the end user making their own adjustments. With the Sun light, there are  many more parameters on both sides. In Explorer, we are currently exposing time-of-day, time zone, latitude, and longitude for the end user to change and affect the sunlight calculation. My question was if we have real (or potential) use cases for exposing all of these parameters or if some of them might be configurable in Story only. I don't have enough knowledge in the content area to make that judgement.

carschno commented:
> Do you have use cases from @kgschoueri or others on that end for the front-facing sun UI parameters?

A use case would be simulation for historic research. In the [Battle at 25 Northumberland Road](https://editions.pure3d.eu/project/1/edition/1/index.html), many details are unknown, including the exact timeline. However, eyewitness reports account for environmental circumstances like sunset in parallel to certain events. Users of the model could visualize different scenarios by setting the time of the day.

carschno commented:
Commit https://github.com/Smithsonian/dpo-voyager/pull/391/commits/4742256b894b3be2b24384a6d66f7c6b8ac240d2 changes the `LightsFollowCam` setting and shows a notification when adding a sunlight because a sunlight should have a fixed position.
Cf. https://github.com/Smithsonian/dpo-voyager/discussions/407 in Q&A.

carschno commented:
I understand we need to clarify if setting date/time/location are actually useful in Story compared to defining them in Tours. @kgschoueri could you help?

BTW, setting these sunlight parameters in Tours also depends on PR https://github.com/Dynamic3D-Project/dpo-voyager/pull/21.

carschno commented:
I have removed the latitude/longitude and the timezone controls from the light tool.

carschno commented:
> In absence of a response from @kgschoueri I've merged this as-is with [rc-59](https://github.com/Smithsonian/dpo-voyager/tree/rc-59). If we hear different in the future we can always remove the sunlight restriction.

That is great, thanks! I agree it is best to proceed iteratively, now that we have the functionality tested and validated. Further fine-tuning would be easy to implement from now on.
